In [2]:
# these warnings are fine. you can ignore them.
from util import *
import numpy as np
import random, math

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("Imports finished.")

/home/niyati/miniconda3/lib/python3.11/site-packages/torch_geometric/typing.py:54: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/niyati/miniconda3/lib/python3.11/site-packages/libpyg.so)
  warnings.warn(f"An issue occurred while importing 'pyg-lib'. "
/home/niyati/miniconda3/lib/python3.11/site-packages/torch_geometric/typing.py:110: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/niyati/miniconda3/lib/python3.11/site-packages/libpyg.so)
  warnings.warn(f"An issue occurred while importing 'torch-sparse'. "


Imports finished.


# Setting Up Dataset/Model/Ground Truth

In [3]:
dataset = Dataset(root='/tmp/Cora', name='Cora', device=device)
data, in_feats, h_feats, num_classes = dataset.get_data()

model = get_model(in_feats, h_feats, num_classes)

ground_truth = get_ground_truth(model, data)
print(ground_truth)

0.749


# Experiments
**Note:** The ground truth here is $0.749$

**Note:** This experiment was done before the development of new functions. The experiments will not work until transferred over to current function.

#### Accidental Degree Experiment

In [4]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()

edge_set = defaultdict(list)
c = 0.1

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

NameError: name 'reset_dataset' is not defined

In [45]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()
class_and_added = {label: [] for label in set(g.ndata["label"].tolist())}

edge_set = defaultdict(list)
c = 0.15

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.

----
The accuracy has not changed.
Change in edges:  41.0


In [46]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()
class_and_added = {label: [] for label in set(g.ndata["label"].tolist())}

edge_set = defaultdict(list)
c = 0.2

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.

----
The accuracy has not changed.
Change in edges:  186.0


In [47]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()
class_and_added = {label: [] for label in set(g.ndata["label"].tolist())}

edge_set = defaultdict(list)
c = 0.25

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.

----
The accuracy has changed by 0.0020
Change in edges:  309.0


In [48]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()
class_and_added = {label: [] for label in set(g.ndata["label"].tolist())}

edge_set = defaultdict(list)
c = 0.3

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.

----
The accuracy has changed by 0.0020
Change in edges:  315.0


In [49]:
from collections import defaultdict

reset_dataset()
modified_graph = cora_dataset[0]
init_edges = modified_graph.number_of_edges()
class_and_added = {label: [] for label in set(g.ndata["label"].tolist())}

edge_set = defaultdict(list)
c = 1

for i in range(0, num_classes):
    for j in range(0, len(homophilic_set[i])):
        edge_set[modified_graph.out_degrees(j)].append(j)


# note this implementation is likely inefficient
for key in edge_set:
    if key == 0:
        continue
    else:
        for elem in edge_set[key]:
            n = min(int(math.floor(key * c)), len(edge_set[key]))
                
            for i in range(0, n):
                b_node = elem
                c_node = random.choice(edge_set[key])

                if b_node == c_node or modified_graph.has_edges_between(b_node, c_node):
                    continue

                modified_graph.add_edges(b_node, c_node)
                modified_graph.add_edges(c_node, b_node)
            

changed_acc(ground_truth, test(modified_graph))
print("Change in edges: ", (modified_graph.number_of_edges() - init_edges)/2)

  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.

----
The accuracy has changed by 0.0050
Change in edges:  1605.0
